README
NAME

    Plack::Middleware::Cache - Caching Reverse Proxy for Plack

SYNOPSIS

    # Note! This modules is work in progress, and the
    # namespace overlaps heavily with the module released
    # to CPAN by Ingy döt Net, so there are no immediate plans
    # to release this module as such.

    my $chi = CHI->new(
        driver => 'File',
        root_dir => 'common/cache',
    );

    enable 'Cache', chi => $chi, rules => [
        qr{^/api/}          => undef,
        qr{\.(jpg|png)$}    => { expires_in => '5 min' },
    ], scrub => [ 'Set-Cookie' ], cachequeries => 1;

DESCRIPTION

    Enable HTTP caching for Plack-based applications.

    Mathing URI's (rules) are cached with the specified
    expiry time / ttl value to the CHI cache.

    Current implementation (on master branch) does not
    support cache validation. See devel branch for work in
    progress towards this.

AUTHOR

    Panu Ervamaa <pnu@cpan.org>

    This module is largely based on Rack::Cache by Ryan Tomayko.
    See http://rtomayko.github.com/rack-cache/ for more information.

LICENCE

    Copyright (c) 2011. Panu Ervamaa.

    This program is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html


Changes
Changelog for Plack::Middleware::Cache::CHI

{{$NEXT}}

0.102     2015-08-25 16:51:04+03:00 Europe/Helsinki
  * Documentation fix

0.101     2015-08-25 16:44:17+03:00 Europe/Helsinki
  * Rename Plack::Middleware::Cache -> Plack::Middleware::Cache::CHI
  * Initial CPAN release


lib/Plack/Middleware/Cache.pm
package Plack::Middleware::Cache;
use strict;
use warnings;
use parent qw/Plack::Middleware/;

use Plack::Util::Accessor qw( storage ttl scrub allow_reload );
use Data::Dumper;
use Plack::Request;
use Plack::Response;
use Time::HiRes qw( gettimeofday );

our @trace;
our $timer_call;
our $timer_pass;

sub _uinterval {
    my ( $t0, $t1 ) = ( @_, [gettimeofday] );
    ($t1->[0] - $t0->[0]) * 1_000_000 + $t1->[1] - $t0->[1];
}

## call: the call hook for the Plack stack.
## Take one parameter (the env) and return a reponse tuple.
##
sub call {
    my ($self,$env) = @_;

    ## Pass-thru streaming responses
    return $self->app->($env)
        if ( ref $env eq 'CODE' );

    ## Localize trace for this request
    local @trace = ();
    local $timer_pass = undef;
    local $timer_call = [gettimeofday];

    my $req = Plack::Request->new($env);
    my $r = $self->handle($req);
    my $res = Plack::Response->new(@$r);

    ## Add trace and cache key to response headers
    $timer_call = _uinterval($timer_call);
    my $trace = join q{, }, @trace;
    my $key = $self->cachekey($req);

    ## The subrequest is timed separately
    if ( $timer_pass ) {
        $timer_call -= $timer_pass;
        $res->headers->push_header(
            'X-Plack-Cache-Time-Pass' => "$timer_pass us",
        );
    }

    $res->headers->push_header(
        'X-Plack-Cache' => $trace,
        'X-Plack-Cache-Key' => $key,
        'X-Plack-Cache-Time' => "$timer_call us",
    );

    $res->finalize;
}

## handle: take request and decide if we should pass it thru,
## try to lookup from cache, or invalidate and pass thru.
## Return the response tuple that should be sent back to the client.
##
sub handle {
    my ($self,$req) = @_;

    if ( $req->method eq 'GET' or $req->method eq 'HEAD' ) {
        if ( $req->headers->header('Expect') ) {
            push @trace, 'expect';
            $self->pass($req);
        } elsif ( $req->cache_control->{'no-cache'} && $self->allow_reload ) {
            push @trace, 'reload';
            $self->fetch($req);
        } else {
            $self->lookup($req);
        }
    } else {
        $self->invalidate($req);
    }
}

## pass: delegate request to the backend and return
## the response. A separate timer is recorded for logging
## purposes.
##
sub pass {
    my ($self,$req) = @_;
    push @trace, 'pass';
    
    $timer_pass = [gettimeofday];
    my $res = $self->app->($req->env);
    $timer_pass = _uinterval($timer_pass);
    
    return $res;
}

## invalidate: remove any matching entries from cache
## storage and pass thru the request.
##
sub invalidate {
    my ($self,$req) = @_;
    push @trace, 'invalidate';
    
    $self->storage->remove( $self->cachekey($req) );
    $self->pass($req);
}

## ttl: determine the configured ttl (or acceptable range of ttl, TODO) for
## a given request. If return value is a positive scalar, it's ttl in seconds,
## negative value means must that the entry referenced by req (that maybe
## rewritten during match processing in this method) must be invalidated from
## cache. If return value is an array ref, the ttl should be adjusted to that
## range [min,max]. Undef in a range means "no limit", or that normal rules
## should be applied.
##
sub ttl {
    my ($self, $req) = @_;

    my $path, $ttl;
    my @rules = @{ $self->rules || [] };
    while ( @rules || return ) {
        my $match = shift @rules;
        $ttl = shift @rules;
        $path = $req->path_info;
        last if 'CODE' eq ref $match ? $match->($path) : $path =~ $match;
    }
    $req->env->{PATH_INFO} = $path;

    return $ttl;
}

## lookup:  try to serve the response from cache. When a matching cache entry is
## found and is fresh, use it as the response without forwarding any
## request to the backend. When a matching cache entry is found but is
## stale, attempt to validate the entry with the backend using conditional
## GET. When no matching cache entry is found, trigger miss processing.
##
sub lookup {
    my ($self, $req) = @_;
    push @trace, 'lookup';

    ## Determine the default ttl (if defined in configuration)
    ## TODO: apply it to the cached response
    my $ttl = $self->ttl($req);
    return $self->invalidate($req) if ( $ttl < 0 );

    ## Get response from the cache storage
    my $entry = $self->get( $req );
    if ( defined $entry ) {
        push @trace, 'hit';
        $self->refurbish($req,$entry,$ttl) ||
        $self->validate($req,$entry,$ttl);
    } else {
        push @trace, 'miss';
        $self->fetch($req,$ttl);
    }
}

## refurbish: Given a storage entry, return the response
## with updated Age header.
##
sub refurbish {
    my ($self, $req, $entry, $ttl) = @_;
    my ($ereq,$eres) = @{$entry};
    push @trace, 'refurbish';

    return unless $self->is_fresh_enough($entry);

    Plack::Util::header_set($eres->[1], 'Age', 0); ## TODO
    $eres;
}

sub is_fresh_enough {
    my ($self, $entry) = @_;
    ## TODO
    return 1;
}

## validate: make a validation request to the backend. If 304 is returned,
## the stored response can be used (with few headers updated from the
## validation response). For other responses, enter it to the storage
## (if cacheable) and return it as is.
##
sub validate {
    my ($self, $req, $entry, $ttl) = @_;
    my ($ereq,$eres) = @{$entry};
    push @trace, 'validate';

    ## Make a new validation request based on the original req
    my $subreq = Plack::Request->new( $req );
    $subreq->method('GET');
    $subreq->headers->header('If-Modified-Since') = $eres->last_modified;

    ## ETags that client has, and etags we have..
    my %req_etag = map { $_ => 1 } split /\s*,\s*/,
        $subreq->headers->header('If-None-Match');
    my %store_etag = map { $_ => 1 } split /\s*,\s*/,
        Plack::Util::header_get($eres->[1], 'ETag');

    ## Any of these etags satisfy our validation request
    $subreq->headers->header( 'If-None-Match' =>
        join ', ', (keys %req_etag, keys %store_etag)
    );

    my $res = $self->pass( $subreq );
    
    if ( $res->[0] == 304 ) {
        push @trace, 'notmodified';
        
        ## Extract the ETags that this response validated
        my %res_etag = map { $_ => 1 } split /\s*,\s*/,
            Plack::Util::header_get($res->[1], 'ETag');

        ## Return the 304 as is if it validated something we don't have
        my $etag = Plack::Util::header_get($res->[1], 'ETag');
        return $res if $etag && $req_etag{$etag} && !$store_etag{$etag};

        ## Copy various caching related headers to the stored response.
        for ( qw( Date Expires Cache-Control ETag Last-Modified ) ) {
            my $val = Plack::Util::header_get( $res->[1], $_ );
            Plack::Util::header_set( $eres->[1], $val );
        }

        ## And return the stored response
        return $eres;
    }

    ## Store the response
    my $response = Plack::Middleware::Cache::Response->new($res);
    $self->set($req,$res,$ttl) if $response->is_cacheable;

    $res;
}

## fetch: The cache missed or a reload is required. Forward the request to the
## backend and determine whether the response should be stored.
##
sub fetch {
    my ($self, $req, $ttl) = @_;
    push @trace, 'fetch';

    ## Make a request based on the original req
    my $subreq = Plack::Request->new( $req );
    $subreq->method('GET');
    $subreq->headers->remove_header('If-Modified-Since','If-None-Match');
  
    my $res = Plack::Middleware::Cache::Response->new( $self->pass($subreq) );

    ## Mark the response as explicitly private if any of the private
    ## request headers are present and the response was not explicitly
    ## declared public.
    $res->cache_control->{private} = undef
        if $self->is_private($req) &&
        exists $res->cache_control->{public};

    # use our own ttl if defined, or use what's provided
    # cache control can disable the default ttl assigment.
    $ttl = $res->is_must_revalidate
        ? $res->ttl
        : $ttl || $res->ttl;

    ## Store the response
    my $response = $res->finalize;
    $self->set($req,$response,$ttl) if $res->is_cacheable;

    $response;
}

sub is_private {
    my ($self, $req) = @_;
        
    for ( @{ $self->private_headers } ) {
        return 1 if $req->headers->header($_);
    }

    return;
}

sub cachekey {
    my ($self, $req) = @_;

    my $uri = $req->uri->canonical;

    $uri->query(undef)
        if not $self->cachequeries;

    $uri->as_string;
}

sub get {
    my ($self, $req) = @_;
    push @trace, 'get';

    my $key = $self->cachekey($req);
    $self->storage->get( $key );
}

sub set {
    my ($self, $req, $res, $ttl) = @_;
    push @trace, 'set';
 
    ## Read in filehandle bodies and scrub the headers before storage
    my $key = $self->cachekey($req);
    my $response = $self->slurp($res);

    $self->storage->set( $key, [$req->headers,$response], $ttl );
}

sub slurp {
    my ($self, $res) = @_;

    foreach ( @{ $self->scrub || [] } ) {
        Plack::Util::header_remove( $res->[1], $_ );
    }

    my $body;
    Plack::Util::foreach( $res->[2], sub {
        $body .= $_[0] if $_[0];
    });

    return [ $res->[0], $res->[1], [$body] ];
}

1;


lib/Plack/Middleware/Cache/Response.pm
package Plack::Middleware::Cache::Response;
#use strict;
#use warnings;
use parent qw(Plack::Response);
use Data::Dumper;
use DateTime;
use DateTime::Format::HTTP

our $VERSION = '0.02';


# Just like Plack::Response::new, but extract the cache_control
# header to a hash first.

sub new {
    my $self = shift->SUPER::new(@_);
    
    my $str = $self->headers->header('Cache-Control');
    my $cc;
    foreach my $kv ( split m{\s*,\s*}, $str ) {
        next if $kv eq '';
        my ($k,$v) = split '=', lc $kv;
        $cc->{$k} = $v;
    }

    $self->{cache_control} = $cc;
    $self->{now} = DateTime->now;
    $self;
}


# Just like Plack::Response::finalize, but bake in the cache_control
# hash values into headers first.

sub finalize {
    my $self = shift;

    my $cc = $self->{cache_control};
    my $str = join ', ', map {
        ($cc->{$_} ? $_.'='.$cc->{$_} : $_)
    } keys %$cc;

    $self->headers->header( 'Cache-Control', $str ) if $str;
    $self->SUPER::finalize;
}


# Get/set the Expires and Date headers as a DateTime object.

sub date { shift->_header_date('Date',@_) }
sub expires { shift->_header_date('Expires',@_) }


# Get/set ETag, Vary and Last-Modified headers as-is
# Note: last_modified is NOT a DateTime object and must
# be compared by string equality vs. If-Modified-Since!

sub etag { shift->headers->header('ETag',@_) }
sub vary { shift->headers->header('Vary',@_) }
sub last_modified { shift->headers->header('Last-Modified',@_) }


# Age of the response in seconds. From Age header, or difference
# between Date header and now.

sub age {
    my $self = shift;
    
    my $age = $self->headers->header('Age',@_);
    return $age if defined $age;
    return 0 if not defined $self->date;
    
    $age = $self->{now}
        ->subtract_datetime_absolute($self->date)
        ->in_units('seconds');

    return $age > 0 ? $age : 0;
}


# The number of seconds after the time specified in the response's Date
# header when the the response should no longer be considered fresh. First
# check for a s-maxage directive, then a max-age directive, and then fall
# back on an expires header; return nil when no maximum age can be
# established.

sub max_age {
    my $self = shift;

    my $max_age = $self->{cache_control}->{'s-maxage'} ||
                  $self->{cache_control}->{'max-age'};
    
    return $max_age if $max_age;

    my $expires = $self->expires;
    return if not defined $expires;

    my $date = $self->date || $self->{now};
    return $expires
        ->subtract_datetime_absolute($date)
        ->in_units('seconds');
}


# The response's time-to-live in seconds, or undef when no freshness
# information is present in the response. When the responses #ttl
# is <= 0, the response may not be served from cache without first
# revalidating with the origin.

# Set the response's time-to-live for shared caches to the specified number
# of seconds. This adjusts the Cache-Control/s-maxage directive.

sub ttl {
    my ($self,$seconds) = @_;
    if ($seconds) {
        $self->{cache_control}->{'s-maxage'} = $self->age + $seconds;
    } else {
        my $max_age = $self->max_age;
        return if not defined $max_age;
        return $max_age - $self->age;
    }
}


# Determine if the response is "fresh". Fresh responses may be served from
# cache without any interaction with the origin. A response is considered
# fresh when it includes a Cache-Control/max-age indicator or Expiration
# header and the calculated age is less than the freshness lifetime.

sub is_fresh {
    my $self = shift;

    my $ttl = $self->ttl;
    return if not defined $ttl;

    return $ttl > 0;
}


# Determine if the response includes headers that can be used to validate
# the response with the origin using a conditional GET request.

sub is_validateable {
    my $self = shift;
    
    return $self->last_modified || $self->etag;
}


# Determine if the response is worth caching under any circumstance. Responses
# marked "private" with an explicit Cache-Control directive are considered
# uncacheable.
#
# Responses with neither a freshness lifetime (Expires, max-age) nor cache
# validator (Last-Modified, ETag) are considered uncacheable.

sub is_cacheable {
    my $self = shift;

    # Status codes of responses that MAY be stored by a cache or used in reply
    # to a subsequent request. http://tools.ietf.org/html/rfc2616#section-13.4
    my @cacheable_codes = qw( 200 203 300 301 302 404 410 ); 

    return if not grep { $_ eq $self->status } @cacheable_codes;
    return if $self->{cache_control}->{'no-store'};
    return if $self->{cache_control}->{'private'};

    return $self->is_validateable || $self->is_fresh;
}


# Indicates that the cache must not serve a stale response in any
# circumstance without first revalidating with the origin. When present,
# the TTL of the response should not be overriden to be greater than
# the value provided by the origin.

sub is_must_revalidate {
    my $self = shift;

    return exists $self->{cache_control}->{'must-revalidate'} ||
           exists $self->{cache_control}->{'proxy-revalidate'};
}


# Mark the response stale by setting the Age header to be equal to the
# maximum age of the response.

sub expire {
    my $self = shift;
    $self->age( $self->max_age ) if $self->is_fresh;
}


# Modify the response so that it conforms to the rules defined for
# '304 Not Modified'. This sets the status, removes the body, and
# discards any headers that MUST NOT be included in 304 responses.
#
# http://tools.ietf.org/html/rfc2616#section-10.3.5

sub make_not_modified {
    my $self = shift;

    # Headers that MUST NOT be included with 304 Not Modified responses.
    # http://tools.ietf.org/html/rfc2616#section-10.3.5
    my @not_modified_omit = qw(
        Allow Content-Encoding Content-Language Content-Length
        Content-MD5 Content-Type Last-Modified
    );
    
    $self->status(304);
    $self->body('');
    $self->headers->remove_header( @not_modified_omit );
}


#######################

# Internal get/set helper method for manipulating date-based headers.
# (Expires and Date)

sub _header_date {
    my ($self,$name,$value) = @_;

    if ($value) {
        my $d = DateTime::Format::HTTP->format_datetime($value);
        return $self->headers->header($name => $d);
    }

    my $d = $self->headers->header($name);
    return if not defined $d;

    DateTime::Format::HTTP->parse_datetime($d);
}


1;


lib/Plack/Middleware/Cache/CHI.pm
use strict;
package Plack::Middleware::Cache::CHI;
# ABSTRACT: Caching Reverse Proxy for Plack

use warnings;
use parent qw/Plack::Middleware/;

use Plack::Util::Accessor qw( chi rules scrub cachequeries trace );
use Data::Dumper;
use Plack::Request;
use Plack::Response;
use Time::HiRes qw( gettimeofday );

# VERSION

our @trace;
our $timer_call;
our $timer_pass;

sub _uinterval {
    my ( $t0, $t1 ) = ( @_, [gettimeofday] );
    ($t1->[0] - $t0->[0]) * 1_000_000 + $t1->[1] - $t0->[1];
}

sub call {
    my ($self,$env) = @_;

    ## Pass-thru streaming responses
    return $self->app->($env)
        if ( ref $env eq 'CODE' );

    ## Localize trace for this request
    local @trace = ();
    local $timer_pass = undef;
    local $timer_call = [gettimeofday];

    my $req = Plack::Request->new($env);
    my $r = $self->handle($req);
    my $res = Plack::Response->new(@$r);

    ## Add trace and cache key to response headers
    $timer_call = _uinterval($timer_call);
    my $trace = join q{, }, @trace;
    my $key = $self->cachekey($req);

    ## The subrequest is timed separately
    if ( $timer_pass ) {
        $timer_call -= $timer_pass;
        $res->headers->push_header(
            'X-Plack-Cache-Time-Pass' => "$timer_pass us",
        );
    }

    $res->headers->push_header(
        'X-Plack-Cache' => $trace,
        'X-Plack-Cache-Key' => $key,
        'X-Plack-Cache-Time' => "$timer_call us",
    );

    $res->finalize;
}

sub handle {
    my ($self,$req) = @_;

    if ( $req->method eq 'GET' or $req->method eq 'HEAD' ) {
        if ( $req->headers->header('Expect') ) {
            push @trace, 'expect';
            $self->pass($req);
        } else {
            $self->lookup($req);
        }
    } else {
        $self->invalidate($req);
    }
}

sub pass {
    my ($self,$req) = @_;
    push @trace, 'pass';
    $timer_pass = [gettimeofday];

    my $res = $self->app->($req->env);

    $timer_pass = _uinterval($timer_pass);
    return $res;
}

sub invalidate {
    my ($self,$req) = @_;
    push @trace, 'invalidate';
    $self->chi->remove( $self->cachekey($req) );
    $self->pass($req);
}

sub match {
    my ($self, $req) = @_;

    my $path;
    my $opts;

    my @rules = @{ $self->rules || [] };
    while ( @rules || return ) {
        my $match = shift @rules;
        $opts = shift @rules;
        $path = $req->path_info;
        last if 'CODE' eq ref $match ? $match->($path) : $path =~ $match;
    }
    return $opts;
}

sub lookup {
    my ($self, $req) = @_;
    push @trace, 'lookup';

    my $opts = $self->match($req);

    return $self->pass($req)
        if not defined $opts;

    return $self->invalidate($req)
        if ( $req->param and not $self->cachequeries );

    my $entry = $self->fetch( $req );
    my $res = [ 500, ['Content-Type','text/plain'], ['ISE'] ];

    if ( defined $entry ) {
        push @trace, 'hit';
        $res = $entry->[1];
        return $self->invalidate($req)
            if not $self->valid($req,$res);
    } else {
        push @trace, 'miss';
        $res = $self->delegate($req);
        $self->store($req,$res,$opts)
            if $self->valid($req,$res);
    }
    return $res;
}

sub valid {
    my ($self, $req, $res) = @_;

    my $res_status = $res->[0];

    return
        unless (
            $res_status == 200 or
            $res_status == 203 or
            $res_status == 300 or
            $res_status == 301 or
            $res_status == 302 or
            $res_status == 404 or
            $res_status == 410
        );

    return 1;
}

sub cachekey {
    my ($self, $req) = @_;

    my $uri = $req->uri->canonical;

    $uri->query(undef)
        if not $self->cachequeries;

    $uri->as_string;
}

sub fetch {
    my ($self, $req) = @_;
    push @trace, 'fetch';

    my $key = $self->cachekey($req);
    $self->chi->get( $key );
}

sub store {
    my ($self, $req, $res, $opts) = @_;
    push @trace, 'store';

    my $key = $self->cachekey($req);
    $self->chi->set( $key, [$req->headers,$res], $opts );
}

sub delegate {
    my ($self, $req, $opts) = @_;
    push @trace, 'delegate';

    my $res = $self->pass($req);
    foreach ( @{ $self->scrub || [] } ) {
        Plack::Util::header_remove( $res->[1], $_ );
    }

    my $body;
    Plack::Util::foreach( $res->[2], sub {
        $body .= $_[0] if $_[0];
    });

    return [ $res->[0], $res->[1], [$body] ];
}

1;

=for test_synopsis use Plack::Builder;

=head1 SYNOPSIS

    my $chi = CHI->new(
        driver => 'File',
        root_dir => 'common/cache',
    );

    enable 'Cache::CHI', chi => $chi, rules => [
        qr{^/api/}          => undef,
        qr{\.(jpg|png)$}    => { expires_in => '5 min' },
    ], scrub => [ 'Set-Cookie' ], cachequeries => 1;

=head1 DESCRIPTION

Enable HTTP caching for Plack-based applications.

Mathing URI's (rules) are cached with the specified
expiry time / ttl value to the CHI cache.

Current implementation (on master branch) does not
support cache validation. See devel branch for work in
progress towards this.

=head1 SEE ALSO

This module is largely based on Rack::Cache by Ryan Tomayko.
See http://rtomayko.github.com/rack-cache/ for more information.

This module was earlier called Plack::Middleware::Cache and available
only thru github because of name conflict with another similar CPAN module.

=cut
